### DISCUSSION TITLE
Node Matching between Graphs

### DISCUSSION THREAD

**Discussion by dieser-max** [pyg-team/pytorch_geometric#7169] [Q&A] [answered]
Hey, I have a problem and need help.

For my research experiment, I have to do some sort of node matching between graphs and I did not find an example that could help me, but I do not think I am the first with this Problem.

My Problem is that I have a few small Graphs with weighted edges. One for each lab run. All those Nodes have quite similar features, but sometimes nodes are missing. To compare my results, I need to match the nodes between the graphs. E.g: In Run/Graph 1 Node 3 corresponds to Node 5 in Graph 2 and is missing in Graph 3 up to 20 Graphs just based on the node embedding in their respective Graphs. The weights of the predicted Links do not matter.

I wanted to solve it by doing a Link prediction between the graphs. In the end, I want to have an adjacency matrix that tells me how the nodes are linked within their graph and between all the others (or just between the Graphs if this is better).

I am currently generating synthetic Data that is based on multiple different runs (I manipulate the node feature values and remove some nodes) to train the model, but I am not sure whether it is better to input the Graphs as pairs or as bigger groups.
If I input them as a big Group I would have an adjacency Matrix where on and near the diagonal I have the edges of Nodes within one Graph and on the sub-diagonals I would have the links between the Graphs. Or is there a better way, maybe with the DataLoader, but I can't think of a way that keeps the Links between the Graphs for Training.
If there is any research, frameworks, or repos with a similar problem, I would be delighted if you told me. Thanks in advance!

**Comment by rusty1s** [comment, marked as answer]
You might find https://github.com/rusty1s/deep-graph-matching-consensus interesting.

**Reply by dieser-max** [reply]
Thank you for your fast answer and the paper. Unfortunately, I did not manage to run the code for the Pascal data by now, but after looking into it, I found that the ValidPairDataset class could be what I need. Can you tell me if I can use the num_classes in the compute_pair function as a representation for node matching?
If I have 2 Graphs where node 1 matches node 1 and node 2 matches node 2 and so on, pairs should return a unit matrix with dim num_nodes x num_nodes. Can I use this to train the dgmc model? And can I simply remove nodes by setting the respective node_featrus and pair_matrix entries to 0?

**Reply by rusty1s** [reply]
Yes, that sounds valid to me.

**Reply by dieser-max** [reply]
Thank you!


~~I am currently working on it but I get the Error: ImportError: 'SplineConv' requires 'torch-spline-conv' when I am running this line~~
~~`conv = SplineConv(in_channels, out_channels, dim, kernel_size=5)`~~

~~But it is installed:~~
~~`torch 2.0.0 pypi_0 pypi`~~

~~`torch-geometric2.3.0 pypi_0 pypi`~~

~~`torch-scatte 2.1.1+pt20cpu pypi_0 pypi`~~

~~`torch-spline-conv 1.2.2+pt20cpu  pypi_0 pypi`~~

~~Do you think I have to use a previous version of torch. pyg and so on. Or is there a fix?~~
Could Fix it!

**Reply by dieser-max** [reply]
But I do have another question regarding the y in the VaildPairDataset class. Since the DGMC Model expects it to be shaped like `[2, num_ground_truths]` and I can't make sens of the num_ground_truths. Because if I insert it like this:


 `'y', tensor([[0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
                  [6, 9, 3, 5, 0, 4, 2, 8, 1, 7]])) `

Meaning 0 of source corresponds to 6 in target. But this does no work. Could you tell me how y has to be shaped to work? Would help me a lot!

**Reply by rusty1s** [reply]
This looks correct to me. What's the error you are seeing?

**Reply by dieser-max** [reply]
Thank you, but I do not get an Error Message if I train the model (currently on a small Data Set). The loss does not decrease and jumps randomly to one out of 4 Values. And the Parameters do not update. However I will try to redefine it since I thought this was because the y dimensions were wrong or displaying a false relationship.

**Reply by dieser-max** [reply]
If I Import 3 Graphs in the DataLoader with 10 nodes and for each Graph and I stack the y's together. so that the dims of y are `torch.Size([2, num_ground_truths])`. Does the [0] line (the index) of y look like this:
´`0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,...`´
or like this:
´`0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 0, 1, 2...`´
The [1] line looks like this:
`3,  4,  1,  2,  5,  9,  8,  6,  7,  0,  3, ...`
Because the DataLoader automatically creates the first version?

**Reply by rusty1s** [reply]
Indices should be incremented, so something like
```
0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,...
```
is how it should look like.

**Reply by dieser-max** [reply]
I do use this. But my Loss does´nt change at all and so does my Acc on train and test. Any idea why this could be?

**Reply by rusty1s** [reply]
This is hard to tell. I suggest to try some popular methods to ensure that there is no trivial bug, e.g., overfitting on a single batch.

**Reply by dieser-max** [reply]
I did try that but since the model did not converge on a one Graph training set and jumped between fixed losses I thought that y had wrong dimensions. And since I tried to copy the pascal_pf model I tried to find my mistake there.

Edit: From my POV I think the problem could be in the comparisso step between the S_L and my y. If my y looks like described above. How can I interpret the S_L and get the loss?

**Reply by dieser-max** [reply]
And is there a solution or workarouund if I have nodes in my source graph, that are not in the target graph?

**Reply by rusty1s** [reply]
The model cannot even fit a single graph? `S_L` is the soft assignment, and `y` are the indices to query as ground-truth correspondences.

> And is there a solution or workarouund if I have nodes in my source graph, that are not in the target graph?

We haven't implemented that yet, but one solution is to let each source node either map to node in the target graph or to a dummy node in the target graph (which would signal a non-existing correspondence).

**Reply by dieser-max** [reply]
Ok. Thank you. But yes the model can´t fit a single Graph and I am sure it´s caused by the way I defined my Data Classes. But I was not able to find a solution. In the test case the model predicts the same class e.g 8 for all nodes in the Batch.

**Reply by rusty1s** [reply]
So I think this is not a batching problem. Do you have a minimal example to reproduce?

**Reply by dieser-max** [reply]
Yes I can provide it, but could it be that Spline has a Problem with my edge featrue which I defined as the inverse Euclidian Distance (values > 1 are possilbe). And do I have to transform my Graphs like this ? 
`transform = T.Compose([T.Constant(), T.KNNGraph(k=8), T.Cartesian(), ])`

Also my S_L tensor only returns 0 or 1, while yours also has other values (pascal-example): `S_L tensor([[0.0895, 0.0679, 0.0645,  ..., 0.0756, 0.0878, 0.0900], ..... ,  grad_fn=<IndexBackward0>)`

**Reply by rusty1s** [reply]
As long as `SplineConv` receives edge features in range `[0, 1]`, this sounds all fine to me.

`S_L` should have continuous values, otherwise it is not differentiable. Does this might be the reason?

**Reply by dieser-max** [reply]
Thank you. I will look into that. But values greater than 1 will be a problem for edge features, right?
Interestingly the S_0 and S_L start with continuous values, but after a few Epochs they take only 0 and 1 as values. But only if I use a big number of layers

**Reply by rusty1s** [reply]
> But values greater than 1 will be a problem for edge features, right?

Only for `SplineConv`.

**Reply by dieser-max** [reply]
> We haven't implemented that yet, but one solution is to let each source node either map to node in the target graph or to a dummy node in the target graph (which would signal a non-existing correspondence).

Does this mean I need to know before the prediction whether one node is in one graph but not in the other one?

**Reply by rusty1s** [reply]
Only during training, during prediction the model should then figure out to map to this "dummy" node.

**Reply by dieser-max** [reply]
I think I have figured out most of the problems I have had by redefining my data structures. But now the Dataloader seems to have an issue with batchting multiple datasets with different node size together. In your examples this error does not occur?

`Sizes of tensors must match except in dimension 0.`

**Reply by rusty1s** [reply]
PyG should be able to handle varying node sizes, but the feature sizes need to be the same. Is this an issue in your case?

**Reply by dieser-max** [reply]
That´s odd, because all nodes have the same number of features. Only the number of nodes in the graph is different.
But the problem could be the way I concatenate the Data_sets the y values in particular.

**Reply by rusty1s** [reply]
Mh, okay. Can you find out for which of the data attributes the error happens exactly?

**Reply by dieser-max** [reply]
Yes, the problem was that I carried an y with a different shape in my Data Object, that I used for a visualization (and now where in the model). I removed it and it works.
Thank you for your help so far!
Could you tell me how I could optimize the model so that it can handle the case that a node is in the target graph and not in the source graph and vice versa? It really seems to struggle there.
And could you tell me your opinion on the case that the testing accuracy (over many tries) is higher than the training accuracy?

**Reply by rusty1s** [reply]
This is a bit tricky because the current code base is not designed for this. The idea would be that you map all non-existing mappings to a dummy node in the reverse graph (by modifying `y`, and by appending a "dummy node" with zero features and no connections to your graph). That should unblock you.
